### DISCUSSION TITLE
Extend PromQL

### DISCUSSION THREAD

**Discussion by codesome** [prometheus/prometheus#9871] [Ideas]
This issue is meant to gather the requirements and allow comments around extending PromQL.

Feel free to comment on this issue with your use cases which you are not able achieve currently with PromQL.

When we identify some use-cases that we want to solve, we can open dedicated GitHub issues for them.

I plan to keep this issue as the main spot for discussions (and not mailing list) as it's easier to link back later.

**Comment by Cho1409** [comment]
Hey @codesome, thank you for the move

I'm struggling with a promql solution right now. I need to catch when there's two k8s deployment running with different builds. Example:
- microservice-build-1 is running
- just deployed the microservice-build-2
- both builds are running, which is bad, and they stay like that for 20-30 minutes, until the old one dies as expected

So how can I setup a promql/alert for that case?

Thank you!

**Comment by benclapp** [comment]
One of the scenarios that is often quite hard to query for is an `increase()` or `delta()` since a specific point in time. 

Some use cases I could imagine would be a range query on a dashboard, showing the cumulative increase over time for a counter. Also count of events since start of day/week/month as an instant query. 

At least with my current understanding, it's not possible to query for cumulative charts as the range selector is always a fixed range for each step in the graph.

**Comment by codesome** [comment]
(Reply to https://github.com/prometheus/prometheus/issues/7878#issuecomment-685967692)

@Cho1409 I think that's already doable with current PromQL. You can do something like `count(some_metric{label=~"<relevant pattern>"}) > 1` where the label would be different for different builds.

**Comment by dykow** [comment]
* Ability to detect which unique metric disappeared and when it happened . 
* Say some exporter exposes `bool_value_metic`, which can be 1 or 0. I want to be able to detect when the value changed say from 1 to 0 and send alert with `labels` of the metric if the value 0 persists for some specific amount of time.

**Comment by codesome** [comment]
(Reply to https://github.com/prometheus/prometheus/issues/7878#issuecomment-686350469)

@dykow 

> I want to be able to detect when the value changed say from 1 to 0 and send alert with labels of the metric if the value 0 persists for some specific amount of time.

That sounds like something which can be done with PromQL for your example. You can use `delta(metric[1m]) < 0`.

Talking about detecting metrics disappearing, you can use `delta` over `count` on that specific metric (now that we have subqueries).

**Comment by codesome** [comment]
Another thing we cannot do natively in PromQL is graphing top N time series. All the details [are in this blog post](https://www.robustperception.io/graph-top-n-time-series-in-grafana).

TL;DR is, if you use `topk` in query_range, you can get much more than `k` time series. We cannot graph just `k` metrics based on some condition (say sum, avg, etc is the highest for past x hours).

While it can be done in Grafana, it would be nice to be able to do it natively in PromQL to write ad-hoc queries.

**Comment by dykow** [comment]
Thanks for Your reply @codesome   :)

> That sounds like something which can be done with PromQL for your example. You can use delta(metric[1m]) < 0.

I can detect that value changed, but I won't be able to keep alert active for longer than 1m (according to Your example query), because after that time delta will be 0 if the value won't change again, so the alert will deactivate.

> Talking about detecting metrics disappearing, you can use delta over count on that specific metric (now that we have subqueries).

`delta` expects `range vector` and `count` expects `instant vector`, so I don't understand how to use them together.
Besides that, this solution won't provide labels, so I will know that some metric with specific name disappeared, but I won't know which exactly.

**Comment by roidelapluie** [comment]
#5094 could be interesting and not possible at the moment.

**Comment by codesome** [comment]
(Reply to https://github.com/prometheus/prometheus/issues/7878#issuecomment-686408449)

@dykow 

> I can detect that value changed, but I won't be able to keep alert active for longer than 1m (according to Your example query), because after that time delta will be 0 if the value won't change again, so the alert will deactivate.

`delta(metric[24h]) < 0`

> delta expects range vector and count expects instant vector, so I don't understand how to use them together.
Besides that, this solution won't provide labels, so I will know that some metric with specific name disappeared, but I won't know which exactly.

`delta(count(your_metric_name)[1h:]) < 0`.

`I won't know which exactly` - right. Can you give a specific use case where you want to use this?

**Comment by dykow** [comment]
> Can you give a specific use case where you want to use this?

@codesome 
I use [vmware_exporter](https://github.com/pryorda/vmware_exporter), which provides metrics given below I want to use for alerting if VM went down (those metrics are the best fit for my taste). I also need the possibility to repeat  the alert until that VM comes up again. (I only want to alert if VM **went down**, not when it was down from the beginning)
```
# HELP vmware_vm_power_state VMWare VM Power state (On / Off)
# TYPE vmware_vm_power_state gauge
vmware_vm_power_state{cluster_name="cluster1",dc_name="dc1",host_name="host1",vm_name="some_vm"} 1.0
vmware_vm_power_state{cluster_name="cluster1",dc_name="dc1",host_name="host1",vm_name="another_vm"} 0.0

# HELP vmware_vm_boot_timestamp_seconds VMWare VM boot time in seconds
# TYPE vmware_vm_boot_timestamp_seconds gauge
vmware_vm_boot_timestamp_seconds{cluster_name="cluster1",dc_name="dc1",host_name="host1",vm_name="some_vm"} 1580118753.414149
```

I can came up with 2 approaches to solve this:
1.  Monitor if `vmware_vm_boot_timestamp_seconds` disappears - this metric is only provided for VMs with `vmware_vm_power_state == 1.0` which corresponds to:
    > delta(count(your_metric_name)[1h:]) < 0

2. Monitor `vmware_vm_power_state` change, which corresponds to:
    > delta(metric[24h]) < 0

**Comment by freeseacher** [comment]
Please add IN operator. recenlty playing with smart data and looking for outliers i found i have to write down 
`metric != 10 and metric != 11 and metric != 17`
that should be relativly simple to implement and usefull

**Comment by belm0** [comment]
> While [topk series in a range] can be done in Grafana, it would be nice to be able to do it natively in PromQL to write ad-hoc queries.

I'll clarify the advantages of having topk_max() et. al. based on practical experience with the implementations in Victoria Metrics:
  * **more efficient** - the datapoints only need to be scanned once.  And in the case where the Grafana panel is below the fold or in a collapsed row, no query happens at all.  Ditto when you are viewing only a single panel and happen to change the time range, by avoiding a flood of unrelated and heavy template queries.
  * **easier to maintain** - rather than maintaining two similar queries (template var and panel queries) which can easily get out of sync, there is only one query.  Additionally, no knowledge of regex variable expansion is required.

**Comment by codesome** [comment]
(Reply to https://github.com/prometheus/prometheus/issues/7878#issuecomment-686834218)

@freeseacher You could do `metric != 10 and metric != 11 and metric != 17` as `metric != 10 != 11 != 17` (Including `=` `<` `>` `<=` `>=` etc in the chain) , I don't see a strong requirement for a IN operator here.

**Comment by freeseacher** [comment]
@codesome yes i can. for me it's a simple spell but quite unbreakable (c)

**Comment by valyala** [comment]
> You could do metric != 10 and metric != 11 and metric != 17 as metric != 10 != 11 != 17

On the other hand `metric in (10, 11, 17)` isn't equivalent to `metric == 10 == 11 == 17`

**Comment by hdost** [comment]
I want to be able to submit a query which looks at the current range data set. To generate a straight line which would then be the min,max,avg across the visible time range.

Use Case: I would like to be able to see average of a particular window.

Data In TSDB
t        Data     Query Result
0       3           5
1       1           5
2       3           5
3       5           5
4       3           5

Possible implementation could be <agg>_over_time(seriesname[-1h:1h]) the -1h annotation meaning a forward looking average.

Not sure if this might have to be something in Grafana or other tools instead.

**Comment by roidelapluie** [comment]
That would be offset -30m instead @hdost

**Comment by codesome** [comment]
Right now negative offsets are not supported but will be added soon I believe. ([Issue to track](https://github.com/prometheus/prometheus/issues/6282))

**Comment by hdost** [comment]
> That would be offset -30m instead @hdost 

As @codesome replied I didn't think they were supported yet.  It I've now followed the issue. 😊 thank you.
